### PR TITLE
Add toDot utility method to SSA control flow graphs

### PR DESCRIPTION
**Pull request by clonker**
formerly implemented in test suite but I found myself needing it also in intermediate steps to visualize the current state of the graph when debugging things

### CODE REVIEW

File: libyul/backends/evm/SSAControlFlowGraph.cpp [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+				},
+				[&](SSACFG::PhiValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::VariableValue const&) -> std::string {
+					return fmt::format("v{}", _var.value);
+				},
+				[&](SSACFG::LiteralValue const& _literal) -> std::string {
+					return formatNumberReadable(_literal.value);
+				}
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.cpp:70]: We could simplify this to:
```suggestion
			GenericVisitor{
				[&](SSACFG::UnreachableValue const&) -> std::string {
					return "[unreachable]";
				},
				[&](SSACFG::LiteralValue const& _literal) -> std::string {
					return formatNumberReadable(_literal.value);
				},
				[&](auto const&) -> std::string {
					return fmt::format("v{}", _var.value);
				}				
```

File: libyul/backends/evm/SSAControlFlowGraph.cpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+	static std::string formatPhi(SSACFG const& _cfg, SSACFG::PhiValue const& _phiValue)
+	{
+		auto const transform = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
+		std::vector<std::string> formattedArgs;
+		formattedArgs.reserve(_phiValue.arguments.size());
+		for (auto const& [arg, entry]: ranges::zip_view(_phiValue.arguments | ranges::views::transform(transform), _cfg.block(_phiValue.block).entries))
+		{
+			formattedArgs.push_back(fmt::format("Block {} => {}", entry.value, arg));
+		}
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.cpp]:
```suggestion
			formattedArgs.push_back(fmt::format("Block {} => {}", entry.value, arg));
```

File: libyul/backends/evm/SSAControlFlowGraph.cpp [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+		if (!formattedArgs.empty())
+			return fmt::format("φ(\\l\\\n\t{}\\l\\\n)", fmt::join(formattedArgs, ",\\l\\\n\t"));
+		else
+			return "φ()";
+	}
+
+	void writeBlock(SSACFG const& _cfg, SSACFG::BlockId const& _id, SSACFG::BasicBlock const& _block, size_t const _functionIndex)
+	{
+		auto const transform = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.cpp]: Maybe `toString` is a better name here?
```suggestion
		auto const toString = [&](SSACFG::ValueId const& valueId) { return varToString(_cfg, valueId); };
```

File: libyul/backends/evm/SSAControlFlowGraph.cpp [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+						label,
+						fmt::join(operation.inputs | ranges::views::transform(transform), ", ")
+					);
+				else
+					m_result << fmt::format(
+						"{} := {}({})\\l\\\n",
+						fmt::join(operation.outputs | ranges::views::transform(transform), ", "),
+						label,
+						fmt::join(operation.inputs | ranges::views::transform(transform), ", ")
+					);
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.cpp]:
```suggestion
				if (!operation.outputs.empty())
					m_result << fmt::format(
						"{} := ",
						fmt::join(operation.outputs | ranges::views::transform(transform), ", ")
					);
				m_result << fmt::format(
					"{}({})\\l\\\n",
					label,
					fmt::join(operation.inputs | ranges::views::transform(transform), ", ")
				);
```

File: libyul/backends/evm/SSAControlFlowGraph.cpp [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+						options += fmt::format("<{0}> {0}", formatNumber(jumpCase.first));
+					}
+					if (!options.empty())
+						options += " | ";
+					options += "<default> default";
+					m_result << fmt::format("Block{}_{}Exit [label=\"{{ JT | {{ {} }} }}\" shape=Mrecord];\n", _functionIndex, _id.value, options);
+					for (auto const& jumpCase: jt.cases)
+					{
+						m_result << fmt::format("Block{}_{}Exit:{} -> Block{}_{};\n", _functionIndex, _id.value, formatNumber(jumpCase.first), _functionIndex, jumpCase.second.value);
+					}
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.cpp]:
```suggestion
						m_result << fmt::format("Block{}_{}Exit:{} -> Block{}_{};\n", _functionIndex, _id.value, formatNumber(jumpCase.first), _functionIndex, jumpCase.second.value);
```

**Review by r0qs — COMMENTED** [review]
Nice! I just have some nitpicks that we ignored when merging in the test, that maybe we could quickly fix, but I'm fine in merging without it as well :)

File: libyul/backends/evm/SSAControlFlowGraph.cpp (hunk 1 above)
Comment (clonker) [reply libyul/backends/evm/SSAControlFlowGraph.cpp:70]: i'd rather keep the code duplication and get a compiler error if we add anything to the variant

**Review by r0qs — APPROVED** [review]

File: libyul/backends/evm/SSAControlFlowGraph.cpp (hunk 1 above)
Comment (r0qs) [reply libyul/backends/evm/SSAControlFlowGraph.cpp:70]: Indeed! Missed that ;)

**Pull request merged**